pylint/config.py

```diff
@@ -79,39 +75,108 @@ def save_results(results, base):
         print('Unable to create file %s: %s' % (data_file, ex), file=sys.stderr)
 
 
-def find_pylintrc():
-    """search the pylint rc file and return its path if it find it, else None
+# TODO: Put into utils
+def walk_up(from_dir):
+    """Walk up a directory tree
+    :param from_dir: The directory to walk up from.
+        This directory is included in the output.
+    :type from_dir: str
+    :returns: Each parent directory
+    :rtype: generator(str)
+    """
+    cur_dir = None
+    new_dir = os.path.expanduser(from_dir)
+    new_dir = os.path.abspath(new_dir)
+
+    # The parent of the root directory is the root directory.
+    # Once we have reached it, we are done.
+    while cur_dir != new_dir:
+        cur_dir = new_dir
+        yield cur_dir
+        new_dir = os.path.abspath(os.path.join(cur_dir, os.pardir))
+
+
+def find_pylintrc_in(search_dir):
+    """Find a pylintrc file in the given directory.
+    :param search_dir: The directory to search.
+    :type search_dir: str
+    :returns: The path to the pylintrc file, if found.
+        Otherwise None.
+    :rtype: str or None
+    """
+    path = None
+
+    search_dir = os.path.expanduser(search_dir)
+    if os.path.isfile(os.path.join(search_dir, 'pylintrc')):
+        path = os.path.join(search_dir, 'pylintrc')
+    elif os.path.isfile(os.path.join(search_dir, '.pylintrc')):
+        path = os.path.join(search_dir, '.pylintrc')
+
+    return path
+
+
+def find_nearby_pylintrc(search_dir=''):
+    """Search for the nearest pylint rc file.
+    :param search_dir: The directory to search.
+    :type search_dir: str
+    :returns: The absolute path to the pylintrc file, if found.
+        Otherwise None
+    :rtype: str or None
     """
-    # is there a pylint rc file in the current directory ?
-    if os.path.exists('pylintrc'):
-        return os.path.abspath('pylintrc')
-    if os.path.exists('.pylintrc'):
-        return os.path.abspath('.pylintrc')
-    if os.path.isfile('__init__.py'):
-        curdir = os.path.abspath(os.getcwd())
-        while os.path.isfile(os.path.join(curdir, '__init__.py')):
-            curdir = os.path.abspath(os.path.join(curdir, '..'))
-            if os.path.isfile(os.path.join(curdir, 'pylintrc')):
-                return os.path.join(curdir, 'pylintrc')
-            if os.path.isfile(os.path.join(curdir, '.pylintrc')):
-                return os.path.join(curdir, '.pylintrc')
-    if 'PYLINTRC' in os.environ and os.path.exists(os.environ['PYLINTRC']):
+    search_dir = os.path.expanduser(search_dir)
+    path = find_pylintrc_in(search_dir)
+
+    for search_dir in walk_up(search_dir):
```

Comment:
Let's check the path condition before going in the loop, something as in:
```
if not path:
   for search_dir in walk_up(search_dir):
      if not os.path.isfile(...):
           break
   ...

if path:
   ...
```

Reply:
Done in an upcoming commit!